bazel-build.el

```diff
@@ -0,0 +1,197 @@
+;;; bazel-build.el --- Emacs utilities for using Bazel
+;; Copyright (C) 2018 Google LLC
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;      http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+;;; Commentary:
+;;
+;; This package provides commands to build and run code using Bazel.
+;; It defines interactive commands bazel-build and bazel-run which
+;; perform completion of available Bazel targets.
+
+;;; -*- lexical-binding: t; -*-
+;;; Code:
+
+(require 'cl-lib)
+
+(defun bazel-build ()
+  "Find and build a Bazel target."
+  (interactive)
+  (funcall-interactively #'bazel-build--run-bazel-command "build"))
+
+(defun bazel-run ()
+  "Find and run a Bazel target."
+  (interactive)
+  (funcall-interactively #'bazel-build--run-bazel-command "run"))
+
+(defun bazel-build--run-bazel-command (command)
+  "Run Bazel tool with given COMMAND, e.g. build or run."
+  (let* ((file-name (buffer-file-name))
+         (workspace-root (bazel-build--find-workspace-root file-name))
+         (initial-input (bazel-build--suggest-initial-input file-name workspace-root))
+         (target (completing-read
+                  (format "bazel %s " command) ; prompt
+                  #'bazel-build--completions   ; collection
+                  nil                          ; predicate
+                  nil                          ; require-match
+                  initial-input)))             ; initial-input
```

Comment:
According to the docstring of `completing-read` *INITIAL-INPUT* is deprecated, so it might be best to not use it, or use *DEF* instead.

Reply:
I'd like to save this for a follow-up PR, since I suspect that will entail moving/refactoring a significant amount of the implemented functionality.

Reply:
That's OK, but please add a TODO then so this is not forgotten.